mtl: drv: Power off flow, HPSRAM memory save/restore to an external buffer

tmleman opened the pull request:
This PR adds a power flows to Intel MTL TLB driver
It allows the driver to save the whole Translation Table
as well as contents of memory banks to a designated address
It is designed for memory content retention in an persistent
storage during low power state

dcpleung commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -693,4 +698,162 @@ static int sys_mm_drv_mm_init(const struct device *dev)
 
-SYS_INIT(sys_mm_drv_mm_init, POST_KERNEL, 0);
+#ifdef CONFIG_PM_DEVICE
+static void sys_mm_save_context(void *storage_buffer)
```
Please do not use the `sys_mm_` namespace for these newly added functions as these are not official APIs.

tmleman replied:
@dcpleung please give me some suggestions on function names if the new ones turn out to be bad.

dcpleung replied:
New names look good.

dcpleung commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -125,6 +125,7 @@
 		tlb: tlb@17e000 {
 			compatible = "intel,adsp-tlb";
 			reg = <0x17e000 0x1000>;
+			label = "HPSRAM_TLB";
```
Actually, the main branch is in the process of removing usage of `label` properties in device tree. So please change the code so this does not require this.

ceolin replied:
@marcinszkudlinski  the project is removing the label property. Just remove it. Look the comment below about how to get the device handle without it.

dcpleung commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
+	switch (action) {
+	case PM_DEVICE_ACTION_FORCE_SUSPEND:
+		if (context_save_buffer != NULL) {
+			sys_mm_save_context(context_save_buffer);
```
Need to use new name.

dcpleung commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
+
+	case PM_DEVICE_ACTION_RESUME:
+		if (context_save_buffer != NULL) {
+			sys_mm_restore_context(context_save_buffer);
```
Need to use new name.

jxstelter commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
 	 */
-	size_t unused_size = CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE -
-			     unused_l2_start_aligned;
+	size_t unused_size = L2_SRAM_BASE + L2_SRAM_SIZE - UNUSED_L2_START_ALIGNED;
```
Here you reverted change for virtual memory Fix. Virtual memory space is much higher than physical one.

jxstelter reviewed (changes requested):
This patch reverts fix:
SHA-1: afb156a1eb217c5b7bdedddb75752c697cae381a

* soc: ace_v1x: unmap entire unused virtual space

This patch corrects the initial unmapping of unused l2 memory
to unmap until the end of virtual address space instead
of unmapping until the end of l2 physical memory.

jxstelter commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
+	uint32_t start_page = (UNUSED_L2_START_ALIGNED -  POINTER_TO_UINT(L2_SRAM_BASE))
+					/ CONFIG_MM_DRV_PAGE_SIZE;
+
+	for (page_idx = start_page; page_idx < L2_SRAM_PAGES_NUM; page_idx++) {
```
This loop will go only in physical memory range while TLB table is for virtual range. Now it will loose entries for virtual addresses higher than L2_SRAM_BASE + (L2_SRAM_PAGES_NUM * CONFIG_MM_DRV_PAGE_SIZE). 
It should go from CONFIG_KERNEL_VM_BASE to (CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE)

marcinszkudlinski replied:
that's not true.
see line 743 - there's a remap of used physical memory pages 1:1 to virtual space, so I don't care about how the physical pages used to be mapped to a virtual space. Its enough to store all physical memory - mapped 1:1 to virtual!

This is much faster than checking all virtual mapping and taking care for regions mapped to several addresses etc.

dcpleung commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -43,7 +46,18 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 			/* FIXME: this value should come from MM */
 			uint32_t hpsram_mask[1] = { 0x3FFFFF };
+#if CONFIG_PM_DEVICE
+			const struct device *dev = device_get_binding("HPSRAM_TLB");
```
We are in the process of getting rid of labels in DTS. So it would be great if this is done using `DEVICE_DT_GET(DT_NODELABEL(node_label))`. Also, there is no `HPSRAM_TLB` in the tree right now. So if we enable `CONFIG_PM_DEVICE`, this won't build at all.

marcinszkudlinski replied:
obsolete, no more PM interface

ceolin replied:
`DEVICE_DT_GET(DT_NODELABEL(tlb))`

andyross reviewed (commented):
Isn't this a layering violation?   The mm_drv driver is a mapping layer.  It tells you what pointers map to which underlying memory.  Copying that elsewhere in preparation for remapping or reuse or whatever is part of the "what you do with that memory" layer, and AFAICS shouldn't really overlap?

Basically why can't this be a separate driver that just uses mm_drv to do its work?  Or a separate layer within the app memory management code?  I guess I don't see why we need to add new APIs for "weird' hardware to what otherwise looks like a very portable MMU interface.

marcinszkudlinski commented:
@andyross yes and no. 
The code is to  be called at the very last moment before the device is switched off. All devices - including memory management, should be turned off at the moment. Even worse when restoring the memory - nothing is started yet. So I want to avoid any unnecessary calls, dependencies etc. and put the save/restore routines in the tlb driver.

Another reason - I really don't care for virtual memory mapping when saving the memory contents - all I need is save all used *physical* memory banks. The knowledge about usage of physical banks is locked in a TLB driver and only there.

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -5,6 +5,9 @@
 #include <zephyr/pm/pm.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+#include <device.h>
```
zephyr/device.h ?

ceolin commented on soc/xtensa/intel_adsp/ace_v1x/power.c:
```diff
@@ -43,7 +46,18 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+				 * just before ace_power_down
+				 */
+				pm_device_action_run(dev, PM_DEVICE_ACTION_FORCE_SUSPEND);
+			}
```
This is not correct. If you want to suspend / resume devices manually you have to use device runtime API.

Now the most important thing, devices are automatically suspended / resumed by the power subsystem if they are not busy or doing / using device runtime pm.

marcinszkudlinski replied:
I was under impression that Zephyr won't call it automatically unless it is enabled for a device:

bool pm_device_runtime_is_enabled(const struct device *dev)
{
	struct pm_device *pm = dev->pm;

	return pm && atomic_test_bit(&pm->flags, PM_DEVICE_FLAG_RUNTIME_ENABLED);
}

Can't let Zephyr do it automatically, it is too specific/sensitive and must be called right before "turning the light off"....
Changing it to specific API

ceolin replied:
so, you have to use device_runtime_pm for this device as I'd suggested. `pm_device_action_run` is low level function, if you are manually suspend/resuming a device you should use device runtime -> https://docs.zephyrproject.org/latest/services/pm/device_runtime.html

ceolin commented on drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:
```diff
@@ -20,7 +20,7 @@
  */
 
-#define DT_DRV_COMPAT intel_adsp_tlb
+#define DT_DRV_COMPAT DT_DRV_COMPAT_ADSP_TLB
```
why this change ?

marcinszkudlinski replied:
because the string name "intel_adsp_tlb" is used twice, in .h and in .c files, so better to keep the string in a macro 
If you disagree I'll change it


mm_drv_intel_adsp_mtl_tlb.h:
/* device id from the device tree */
#define DT_DRV_COMPAT_ADSP_TLB intel_adsp_tlb

/* size of TLB table */
#define TLB_SIZE DT_REG_SIZE_BY_IDX(DT_INST(0, **DT_DRV_COMPAT_ADSP_TLB**), 0)


mm_drv_intel_adsp_mtl_tlb.c:
#define DT_DRV_COMPAT **DT_DRV_COMPAT_ADSP_TLB**

ceolin reviewed (changes requested):
The pm flow is not correct.

ceolin reviewed (dismissed):
minor issue with the "label" property to be fixed. Other than that +1

dcpleung commented:
Please rebase first.

edersondisouza commented on include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h:
```diff
@@ -1,21 +1,40 @@
+/* maximum size of buffer required for memory save */
+#define TLB_MAX_MEMORY_SAVE_BUFFER_SIZE \
+	(L2_SRAM_SIZE + TLB_SIZE + L2_SRAM_PAGES_NUM * sizeof(void *))
+
```
I wonder if all those defines should actually belong to an "internal" `drivers/mm/mm_drv_intel_adsp_mtl_tlb.h` and only the API bits here.

edersondisouza commented on include/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h:
```diff
@@ -0,0 +1,21 @@
+#define TLB_SIZE DT_REG_SIZE_BY_IDX(DT_INST(0, DT_DRV_COMPAT_ADSP_TLB), 0)
+
+/* numer of L2 memory pages */
+#define L2_SRAM_PAGES_NUM (L2_SRAM_SIZE / CONFIG_MM_DRV_PAGE_SIZE)
```
Not sure if adding unused - in this patch - definitions helps with clean-up. Maybe defer adding those to when they are needed?

marcinszkudlinski replied:
changed

edersondisouza reviewed (dismissed):
Is this PR still relevant? It hasn't been updated for a while...
Edit: Err, just noticed the last comment from 7 days ago - guess I needed to refresh my tab first =D

kv2019i commented:
@marcinszkudlinski working on an update. Changing status to draft

marcinszkudlinski committed:
mtl: drivers: tlb driver cleanup

a small cleanup in the code, preaparation for memory context save

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>
Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

marcinszkudlinski committed:
mtl: soc: add TLB driver context save to platform PowerManagement

Connect the TLB context save to MTL suspend power flow

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
mtl: soc: add context save and restore flow

Context save is saving whole memory to persistent
memory area, than turning off memory and CPU
Context restore is a modified boot flow, where
the previously saved context is restored

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
mtl: soc: TLB driver provides required size of L3 storage buffer

TLB driver knows the required number of bytes for HPSRAM
storage during power off state.

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski commented:
D3 power flow for ACE soc added. Please review again

edersondisouza commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -218,6 +220,7 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 			imr_layout->imr_state.header.imr_restore_vector =
 					(void *)rom_entry_d3_restore;
 			imr_layout->imr_state.header.imr_ram_storage = global_imr_ram_storage;
+
```
Unrelated to this patch - maybe belong to previous one?

marcinszkudlinski committed:
mtl: soc: store power gating state in D3 state

Power gating register must be stored when CPU is in
power off state

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

tmleman reviewed (commented):
LGTM 👍, can't approve since it's my PR

marcinszkudlinski committed:
mtl: dts: add L3 memory definitions macros

Add helper macros for l3 memory definitions from
the Device Tree

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski commented:
Last (I hope...) version. 
Changes since last review: 
 - compilation fixes 
 - cleaner and little bit updates definitions in adsp_memory.h

aborisovich reviewed (dismissed):
looks ready to me, lets merge

stephanosio commented:
@ceolin can you please take a look?

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -148,21 +182,78 @@ __asm__(".align 4\n\t"
+			const struct device *tlb_dev = DEVICE_DT_GET(DT_NODELABEL(tlb));
 
-			power_down(true, uncache_to_cache(&hpsram_mask[0]),
+			__ASSERT_NO_MSG(dev != NULL);
```
Shouldn't it be `tlb_device` ?

marcinszkudlinski replied:
True! But that means - __ASSERT_NO_MSG macro is off and does not check anything

kv2019i replied:
@marcinszkudlinski Ack, work in progress to make asserts enabled in at least subset of our CI -> https://github.com/thesofproject/sof/pull/6530 -- first attempt failed, but will wbe looking at this again.

ceolin commented:
Minor comment, other than that looks good.

marcinszkudlinski commented:
Done, here we go again with (re) approvals...

marcinszkudlinski commented:
The PR was approved previously couple of times. Please go ahead and merge it

aborisovich reviewed (dismissed):
Just do it

marc-hb commented:
commit c929bbcc58c545 fails like this:

```
sof/src/platform/intel/ace/lib/pm_runtime.c
In file included from zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm/ace_v1x-regs.h:11,
                 from sof/src/platform/intel/ace/lib/pm_runtime.c:12:
zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm/adsp_memory.h:43: error: "IMR_BOOT_LDR_MANIFEST_BASE" redefined [-Werror]
   43 | #define IMR_BOOT_LDR_MANIFEST_BASE      (L3_MEM_BASE_ADDR + IMR_BOOT_LDR_MANIFEST_OFFSET)
      | 
In file included from sof/zephyr/../src/../xtos/include/sof/lib/memory.h:11,
                 from sof/zephyr/../src/include/sof/sof.h:13,
                 from sof/zephyr/../src/../xtos/include/sof/lib/pm_runtime.h:19,
                 from sof/src/platform/intel/ace/lib/pm_runtime.c:5:
sof/zephyr/../src/platform/meteorlake/include/platform/lib/memory.h:52: note: this is the location of the previous definition
   52 | #define IMR_BOOT_LDR_MANIFEST_BASE      0xA1042000
      | 
cc1: all warnings being treated as errors
```

Commit 84c09433b574 fails like this:

```
In file included from zephyr/soc/xtensa/intel_adsp/common/include/cpu_init.h:9,
                 from zephyr/soc/xtensa/intel_adsp/ace/power.c:9:
zephyr/soc/xtensa/intel_adsp/ace/power.c: In function 'pm_state_set':
zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm/adsp_memory.h:18:22: error: passing argument 2 of 'memcpy' makes pointer from integer without a cast [-Werror=int-conversion]
   18 | #define LP_SRAM_BASE (DT_REG_ADDR(DT_NODELABEL(sram1)))
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                      |
      |                      long long int
zephyr/soc/xtensa/intel_adsp/ace/power.c:236:56: note: in expansion of macro 'LP_SRAM_BASE'
  236 |                         memcpy(global_imr_ram_storage, LP_SRAM_BASE, LP_SRAM_SIZE);
      |                                                        ^~~~~~~~~~~~
In file included from zephyr/soc/xtensa/intel_adsp/common/include/soc.h:9,
                 from zephyr/include/zephyr/arch/xtensa/irq.h:60,
                 from zephyr/include/zephyr/arch/xtensa/arch.h:27,
                 from zephyr/include/zephyr/arch/cpu.h:27,
                 from zephyr/include/zephyr/kernel_includes.h:33,
                 from zephyr/include/zephyr/kernel.h:17,
                 from zephyr/soc/xtensa/intel_adsp/ace/power.c:6:
zephyr/lib/libc/minimal/include/string.h:42:63: note: expected 'const void * restrict' but argument is of type 'long long int'
   42 | extern void  *memcpy(void *ZRESTRICT d, const void *ZRESTRICT s,
      |                                         ~~~~~~~~~~~~~~~~~~~~~~^
zephyr/soc/xtensa/intel_adsp/ace/power.c: At top level:
zephyr/soc/xtensa/intel_adsp/ace/power.c:168:9: error: '_restore_core_context' is static but used in inline function 'power_off_exit' which is not static [-Werror]
  168 |         _restore_core_context();
      |         ^~~~~~~~~~~~~~~~~~~~~
zephyr/soc/xtensa/intel_adsp/ace/power.c: In function 'pm_state_imr_restore':
zephyr/soc/xtensa/intel_adsp/ace/power.c:198:1: error: 'noreturn' function does return [-Werror]
  198 | }
      | ^
cc1: all warnings being treated as errors

```

marc-hb commented:
Just found https://github.com/thesofproject/sof/pull/6613, looks like a mutual dependency. Thank @marcinszkudlinski for linking the two.
